tracer/src/Datadog.Trace/Telemetry/Transports/JsonTelemetryTransport.cs

```diff
@@ -0,0 +1,94 @@
+    {
+        NullValueHandling = NullValueHandling.Ignore,
+        ContractResolver = new DefaultContractResolver { NamingStrategy = new SnakeCaseNamingStrategy(), }
+    };
+
+    private readonly IApiRequestFactory _requestFactory;
+    private readonly Uri _endpoint;
+
+    public JsonTelemetryTransport(IApiRequestFactory requestFactory, Uri baseEndpoint)
+    {
+        _requestFactory = requestFactory;
+        _endpoint = new Uri(baseEndpoint, TelemetryConstants.TelemetryPath);
+    }
+
+    public async Task<TelemetryPushResult> PushTelemetry(TelemetryData data)
+    {
+        try
+        {
+            // have to buffer in memory so we know the content length
+            var serializedData = SerializeTelemetry(data);
+            var bytes = Encoding.UTF8.GetBytes(serializedData);
+
+            var request = _requestFactory.Create(_endpoint);
+
+            request.AddHeader(TelemetryConstants.ApiVersionHeader, data.ApiVersion);
+            request.AddHeader(TelemetryConstants.RequestTypeHeader, data.RequestType);
+
+            using var response = await request.PostAsync(new ArraySegment<byte>(bytes), "application/json").ConfigureAwait(false);
+            if (response.StatusCode is >= 200 and < 300)
+            {
+                Log.Debug("Telemetry sent successfully");
+                return TelemetryPushResult.Success;
+            }
+            else if (response.StatusCode == 404)
+            {
+                Log.Debug("Error sending telemetry: 404. Disabling further telemetry, as endpoint not found");
+                return TelemetryPushResult.FatalError;
+            }
+            else
+            {
+                Log.Debug<int>("Error sending telemetry {StatusCode}", response.StatusCode);
+                return TelemetryPushResult.TransientFailure;
+            }
+        }
+        catch (Exception ex) when (IsFatalException(ex))
+        {
+            Log.Warning(ex, "Error sending telemetry data, unable to communicate with endpoint");
+            return TelemetryPushResult.FatalError;
+        }
+        catch (Exception ex)
+        {
+            Log.Warning(ex, "Error sending telemetry data");
+            return TelemetryPushResult.TransientFailure;
+        }
+    }
+
+    // Internal for testing
+    internal static string SerializeTelemetry(TelemetryData data)
+    {
+        return JsonConvert.SerializeObject(data, Formatting.None, SerializerSettings);
+    }
+
+    private static bool IsFatalException(Exception ex)
+    {
+        return ex is SocketException
+#if !NETFRAMEWORK
+                   or WebException { InnerException: System.Net.Http.HttpRequestException { InnerException: SocketException } }
+                   or System.Net.Http.HttpRequestException { InnerException: SocketException }
+#endif
+                   or WebException { Response: HttpWebResponse { StatusCode: HttpStatusCode.NotFound } }
+                   or WebException { InnerException: SocketException };
```

Comment:
Pretty succinct with pattern matching! Although for most of these it seems like we could generalize by looking at the `InnerException`s all the way down until we find a `SocketException` (or not). We wouldn't even need the `#if !NETFRAMEWORK` anymore.

Something like:
```csharp
if (ex is WebException { Response: HttpWebResponse { StatusCode: HttpStatusCode.NotFound } })
{
    return true;
}

while (ex != null)
{
    if (ex is SocketException)
    {
        return true;
    }
    
    ex = ex.InnerException;
}

return false;
```

Reply:
Nice 👍 I can probably extract this to a separate helper and reuse it in the other transports too (traces, CI)